data/trie/patriciaMerkleTrie.go

```diff
@@ -445,37 +449,71 @@ func (tr *patriciaMerkleTrie) ResetOldHashes() [][]byte {
 	return oldHashes
 }
 
-// Snapshot creates a new database in which the current state of the trie is saved.
-// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
-func (tr *patriciaMerkleTrie) Snapshot() error {
+// Checkpoint adds the current state of the trie to the snapshot database
+func (tr *patriciaMerkleTrie) Checkpoint() error {
 	tr.mutOperation.Lock()
 	defer tr.mutOperation.Unlock()
 
+	if tr.root.isDirty() {
+		return ErrTrieNotCommited
+	}
+
 	if tr.snapshotInProgress {
 		return ErrSnapshotInProgress
 	}
 	tr.snapshotInProgress = true
 
-	err := tr.root.commit(false, 0, tr.db, tr.db)
+	if len(tr.snapshots) == 0 {
+		_, err := tr.newSnapshotDb()
+		if err != nil {
+			tr.snapshotInProgress = false
+			return err
+		}
+	}
+
+	lastSnapshotId := len(tr.snapshots) - 1
+	err := tr.root.commit(true, 0, tr.db, tr.snapshots[lastSnapshotId])
 	if err != nil {
+		tr.snapshotInProgress = false
 		return err
 	}
 
+	keys := tr.pruningBuffer
+	tr.pruningBuffer = make([][]byte, 0)
+	tr.snapshotInProgress = false
+
+	for i := range keys {
+		err = tr.removeFromDb(keys[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Snapshot creates a new database in which the current state of the trie is saved.
+// If the maximum number of snapshots has been reached, the oldest snapshot is removed.
+func (tr *patriciaMerkleTrie) Snapshot() error {
+	tr.mutOperation.Lock()
+	defer tr.mutOperation.Unlock()
+
+	if tr.root.isDirty() {
+		return ErrTrieNotCommited
+	}
+
+	if tr.snapshotInProgress {
+		return ErrSnapshotInProgress
+	}
+	tr.snapshotInProgress = true
+
 	rootHash := tr.root.getHash()
 
-	db, err := storageUnit.NewDB(
-		storageUnit.DBType(tr.snapshotDbCfg.Type),
-		path.Join(tr.snapshotDbCfg.FilePath, strconv.Itoa(tr.snapshotId)),
-		tr.snapshotDbCfg.BatchDelaySeconds,
-		tr.snapshotDbCfg.MaxBatchSize,
-		tr.snapshotDbCfg.MaxOpenFiles,
-	)
+	db, err := tr.newSnapshotDb()
 	if err != nil {
+		tr.snapshotInProgress = false
 		return err
 	}
-	tr.snapshotId++
-
-	tr.snapshots = append(tr.snapshots, db)
 
 	if len(tr.snapshots) > maxSnapshots {
 		dbUniqueId := strconv.Itoa(tr.snapshotId - len(tr.snapshots))
```

Comment:
L522-523 you remove an old snapshot. Why not calling Close before erasing physical folder?

Reply:
Changed the interface that the snapshots implement (from DBWriteCacher to Persister), and used the Close() method. The interface change was needed because the trie db also uses DBWriteCacher, but the trie db is instantiated as a storage unit, and the snapshots as Persisters. The storage unit interface has DestroyUnit() whereas persister has Close(), so I couldn't add a new method to the DBWriteCacer interface and also expect to implement bothPersister and StorageUnit interfaces.

Reply:
Great :+1: